fix(api): `substitute` calls `on_bytes` after modifying buffer

noib3 opened the pull request:
Problem: `extmark_splice()` was being called before `ml_replace()`, which caused the on_bytes callback to be invoked with the old buffer text instead of the new text.

Solution: store metadata for each match in a kvec, call `ml_replace()` once to update the buffer, then call `extmark_splice()` once per match.

Closes https://github.com/neovim/neovim/issues/36370.

<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

noib3 committed:
fix(api): `substitute` calls `on_bytes` after modifying buffer

Problem: `extmark_splice()` was being called before `ml_replace()`,
which caused the on_bytes callback to be invoked with the old buffer
text instead of the new text.

Solution: store metadata for each match in a growing array, call
`ml_replace()` once to update the buffer, then call `extmark_splice()`
once per match.

Closes https://github.com/neovim/neovim/issues/36370.

glepnir commented on src/nvim/ex_cmds.c:
```diff
@@ -3771,6 +3772,25 @@ static int do_sub(exarg_T *eap, const proftime_T timeout, const int cmdpreview_n
+        linenr_T lnum_after;   // Line number after this substitution
+      } MatchExtmarkData;
+
+      garray_T match_extmarks = GA_EMPTY_INIT_VALUE;
```
prefer `kvec`

noib3 replied:
Done.

justinmk commented on test/functional/lua/buffer_updates_spec.lua:
```diff
@@ -1250,6 +1250,224 @@ describe('lua: nvim_buf_attach on_bytes', function()
       }
     end)
 
+    it('on_bytes sees modified buffer during substitute', function()
```
?? Why should this be supported? We should not be emitting buffer updates during the "live preview". Only after the substitute is confirmed. The "live preview" is NOT an edit of any kind, it is just a preview.

If a plugin wants to do stuff during the preview it should implement an 'inccommand' preview handler (`:help :command-preview`).

I don't see anything like that in these tests so I assume that is not the case here.

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -3771,6 +3771,24 @@ static int do_sub(exarg_T *eap, const proftime_T timeout, const int cmdpreview_n
+      // Track where substitutions started (set once per line).
+      linenr_T lnum_start = 0;
+
+      // Track extmark_splice data for each match in the line.
```
```suggestion
      // Track per-line data for each match.
      // Will be sent as a batch to `extmark_splice` after the substitution is done.
```

justinmk commented on test/functional/lua/buffer_updates_spec.lua:
```diff
@@ -1250,6 +1250,224 @@ describe('lua: nvim_buf_attach on_bytes', function()
       }
     end)
 
+    it('on_bytes sees modified buffer during substitute', function()
```
Preview is performed by calling `do_sub` (which this PR modifies) and then undoing it later (after the user "confirms" the substitute).

It looks like `extmark_splice` (buffer updates) is already called during 'inccommand' preview. I suppose this could be useful for API clients like vscode-neovim which need to "sync" buffers, but I don't think it makes sense to support...

I guess it's orthogonal to the change in this PR.

noib3 commented:
Renamed `MatchExtmarkData` to `LineData`, renamed first test from `on_bytes sees modified buffer during substitute` to `on_bytes sees modified buffer after substitute` to address https://github.com/neovim/neovim/pull/36487#discussion_r2532441448, and added a [new test](https://github.com/neovim/neovim/pull/36487/files#diff-cc8ef97bacb1e75406844f69dc5b2cb115b357ea0218e0953ab7ed5d3fe6e9e1R1471-R1497) to make sure no buffer update event is emitted while the substitution patterns are being typed out.
